### PR TITLE
Add message for character shooting a gun, scales with volume

### LINKED ISSUE #57437: Game doesn't print message about ranged weapon sounds
### Describe the bug

Firing a noisy ranged weapon should be accompanied by message in the log, according to the `item::gun_noise` in `ranged.cpp`, which has a variety of sounds for different types of weapons:

https://github.com/CleverRaven/Cataclysm-DDA/blob/c2e3aafc7a13e2059e4a23142dfc89b8b8e72923/src/ranged.cpp#L1928-L1982

Despite that, no messages appear in the log.

### Steps to reproduce

- Get yourself a crossbow and a bolt.
- Fire crossbow.
- Observe no message in the log.

### Expected behavior

Messages appear in the log when firing from weapons.

### Screenshots

изображение

I fired from V29 laser pistol. Since it has the `LASER` flag, according to the code, the sound should be `Fzzt!`. No message appear at all.

### Versions and configuration

- Windows 10 x64;
- 0.F-8087;
- Tiles;
- Default mods only.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by stubkan**
#### Summary
Interface "Add message for character shooting a gun, scales with volume"

#### Purpose of change

There is no message indicator when you fire a gun, so if you dont see the hit or for some any reason don't recieve a hit confirmation - you are unsure if you fired a gun.  Also its satisfying to read that you're shooting things.

So there is a message now that will play your gunshot at the suitable volume when you fire.

Fixes #57437

#### Testing

You shoot and you score!

image

Tested with various guns of varying sizes - burst fire and single fire, the messages change accordingly.

### CODE REVIEW

**Comment by Night-Pryanik** [comment]
See #57437.

File: src/ranged.cpp [hunk 1]
```diff
@@ -815,6 +815,10 @@ int Character::fire_gun( const tripoint &target, int shots, item &gun )
         const vehicle *in_veh = has_effect( effect_on_roof ) ? veh_pointer_or_null( here.veh_at(
                                     pos() ) ) : nullptr;
 
+        // Add gunshot noise if player is the one who shot
+        const item::sound_data gunnoise = gun.gun_noise( shots > 1 );
+        add_msg_if_player( _( "You shoot. %s" ), gunnoise.sound );
```
Comment (dseguin) [inline src/ranged.cpp:820]:
```suggestion
        // Add gunshot noise if player is the one who shot
        const item::sound_data gunnoise = gun.gun_noise( shots > 1 );
        Creature *target_critter = get_creature_tracker().creature_at( aim );
        const std::string target_str = !!target_critter ? target_critter->disp_name() :
                                       get_map().disp_name( aim );
        add_msg_if_player( _( "You shoot %1$s with your %2$s. %3$s" ), target_str, gun.tname(),
                           uppercase_first_letter( gunnoise.sound ) );
```
This is just an optional suggestion. Just adds a few more details to the message:

shoot_msg

Comment (stubkan) [reply src/ranged.cpp:820]: Oh yeah, thats nicer

Comment (dseguin) [reply src/ranged.cpp:820]: You can also just use `at` in the message:
```
"You shoot at %1$s with your %2$s. %3$s"
```

Comment (stubkan) [reply src/ranged.cpp:820]: What if youre aiming at nothing though?

Comment (dseguin) [reply src/ranged.cpp:820]:
```
get_map().disp_name( aim );
```
This part uses the name for the terrain or furniture at the location if there's no creature there

Comment (stubkan) [reply src/ranged.cpp]: Well, it seems like it's getting a little too complicated at that point, more code and longer lines to read...  I think its good to keep it simple myself

**Comment by Night-Pryanik** [comment]
What if you're firing a crossbow? A laser gun? A flamethrower? A slingshot? The same "bang"?

**Comment by Nerezza** [comment]
This used to be a thing, I'm not sure when or why it was removed. Crossbows, bows, flamethrowers, and lasers also had unique onomonopia for their lines too.

**Comment by stubkan** [comment]
> What if you're firing a crossbow? A laser gun? A flamethrower? A slingshot? The same "bang"?

Oh yeah... those are silent weapons, they wont have a noise...  I should test fire one of those and see if it crashes

**Comment by stubkan** [comment]
> This used to be a thing, I'm not sure when or why it was removed. Crossbows, bows, flamethrowers, and lasers also had unique onomonopia for their lines too.

They are still in

        // Rocket launchers and flamethrowers
        return { 4, _( "Fwoosh!" ) };
    if( fx.count( "LASER" ) || fx.count( "PLASMA" ) ) {
        if( noise < 20 ) {
            return { noise, _( "Fzzt!" ) };
    } else if( fx.count( "WHIP" ) ) {
        return { noise, _( "Crack!" ) };
etc

Silent weapons (crossbows etc) will return nothing
    return { 0, "" }; // silent weapons

So you'll just see "you shoot your crossbow" and nothing else..  Should be good

**Comment by Night-Pryanik** [comment]
Neither crossbows nor laser gun nor flamethrower are silent weapons. Crossbows might be quiet, but not silent.

There ARE sounds of weapons firing in the game code. The issue is that for some reason these sounds ain't printed in game log, as I described in #57437. And thus I'm afraid your solution isn't right, as it doesn't fix the underlying bug.

**Comment by stubkan** [comment]
item::gun_noise is where gun noise messages come from, If there is no entry for a crossbow then it won't have one, unless there is one written for it now.  I think crossbows not having a noise message doesn't really have anything to do with the bug of guns not making noise messages.

They have a noise lvl of about 10, guns are 25...   The noise production is a different function than the gunshot text noise messages that this PR is dealing with..

Although I can make a separate PR later to add crossbow noise messages or other weapons that need one

**Comment by stubkan** [comment]
image

File: src/ranged.cpp [hunk 2]
```diff
@@ -815,6 +815,11 @@ int Character::fire_gun( const tripoint &target, int shots, item &gun )
         const vehicle *in_veh = has_effect( effect_on_roof ) ? veh_pointer_or_null( here.veh_at(
                                     pos() ) ) : nullptr;
 
+        // Add gunshot noise if player is the one who shot
+        const item::sound_data gunnoise = gun.gun_noise( shots > 1 );
+        add_msg_if_player( _( "You shoot your %1$s. %2$s" ), gun.tname(),
+                           uppercase_first_letter( gunnoise.sound ) );
```
Comment (Night-Pryanik) [inline src/ranged.cpp:821]: I'd remove the `You shoot your X` part.
First, I know that I'm firing right now. Second, I know what gun I'm firing right now. Having this in mind, `You shoot your X` part looks superfluous.

```suggestion
        add_msg_if_player( _( "%s" ), uppercase_first_letter( gunnoise.sound ) );
```

**Comment by Night-Pryanik** [comment]
I suggest moving print messages about gun firing into a `make_gun_sound_effect` function.

**Comment by catdach** [comment]
Not the throw more fuel on this fire, but it may be pertinent to remember that we have suppressors in the game. Some of the quieter guns effectively become silent with a suppressor (like the Marlin 39A). I would personally be very confused if the game told me my 0 noise gun was going "bang!"

Even if it wasn't completely silent I would perhaps expect it to make a slightly _different_ sound (thought **_that_** might be a project for another time)?

File: src/ranged.cpp (hunk 2 above)
Comment (stubkan) [reply src/ranged.cpp:821]: It seems to me that some message telling me what I did is standard in this game.

I pick this up. I open that door. I wield that.  By that same logic, should all of those be removed?

I kind of think it should stay in that it says Im shooting in some way, along with the noise effect.

Looking at the screenshots above, the shoot message fits perfectly (in my opinion)

**Comment by stubkan** [comment]
> I suggest moving print messages about gun firing into a `make_gun_sound_effect` function.

I'll have a look into that in a bit.

> Not the throw more fuel on this fire, but it may be pertinent to remember that we have suppressors in the game. Some of the quieter guns effectively become silent with a suppressor (like the Marlin 39A). I would personally be very confused if the game told me my 0 noise gun was going "bang!"
> 
> Even if it wasn't completely silent I would perhaps expect it to make a slightly _different_ sound (thought **_that_** might be a project for another time)?

No fuel on the fire, discussion is good...

The gun onomatopoeia is fetched according to loudness of the weapon.  So it should hopefully scale accordingly with how loud it is.  However, I'll test it with the suppressor you mentioned.

**Comment by stubkan** [comment]
> I suggest moving print messages about gun firing into a `make_gun_sound_effect` function.

Its in there now

> Silent weapons have no noise message when you shoot them

image

File: src/ranged.cpp (hunk 2 above)
Comment (stubkan) [reply src/ranged.cpp]: Also if you shoot a silenced weapon, you won't get any noise message so you should still get some kind of message for that, I think

**Pull request merged**